### ISSUE TITLE
Different image shape used for inference in Pytorch weights and ONNX

### ISSUE THREAD

**Issue by Ian-Work-AI** [ultralytics/ultralytics#19238]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I fine-tuned a model using `imgsz = 640`.

When running inference with model.predict(), I also set `imgsz = 640`, but the terminal output shows that the .pth format used an input shape of [384, 640], while the ONNX format used [640, 640]. This resulted in different predictions.

Which input size is correct, and why do the two format use different sizes even though I set `imgsz = 640`?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ian-Work-AI, thank you for bringing this up and for your interest in Ultralytics 🚀! 

We recommend checking out our [Docs](https://docs.ultralytics.com/) where you can find detailed explanations about [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, as well as common troubleshooting tips. For your specific question, the [Export Guide](https://docs.ultralytics.com/modes/export/) may offer insights into format-specific behavior.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug the issue. Specifically, including a snippet of your code that reproduces the disparity between the `.pth` and ONNX predictions would be greatly beneficial.

If this is a ❓ Question about custom training or inference settings, please provide as much information as possible:
- The exact command or code you are running
- Environment details (e.g., Python, PyTorch, CUDA versions)
- Screenshots or logs of the terminal outputs showing the shape discrepancy

Additionally, verify that your setup adheres to the recommendations in our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure consistency across formats.

To ensure you're running the latest and greatest, please upgrade your `ultralytics` package and all necessary dependencies:

```bash
pip install -U ultralytics
```

We also encourage you to experiment with several verified environments designed for seamless YOLO operation. For a quick and easy way to reproduce results in a controlled setup, try running in one of these environments:
- <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>
- <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

For real-time assistance, connect with our community on [Discord](https://discord.com/invite/ultralytics) 🎧, participate in discussions on [Discourse](https://community.ultralytics.com/), or join the conversation on our [Subreddit](https://reddit.com/r/Ultralytics).

Lastly, an Ultralytics engineer will be here shortly to assist you further 😊. Stay tuned!

**Comment by glenn-jocher** [comment]
@Ian-Work-AI the input shape difference arises from how PyTorch and ONNX handle aspect ratio preservation. PyTorch automatically pads images to maintain aspect ratio during inference, while ONNX uses strict resizing. To ensure consistency, explicitly set `imgsz=[640, 640]` for both formats. This forces square input matching ONNX's behavior. For implementation details, see our [Autobackend documentation](https://docs.ultralytics.com/reference/nn/autobackend).

**Comment by Y-T-G** [comment]
You can read this:

https://github.com/ultralytics/ultralytics/issues/16339#issuecomment-2358362884

**Comment by Ian-Work-AI** [comment]
@glenn-jocher @Y-T-G 

I think my question is not about how to resize the image to [640, 640].

Let me elaborate on why I’m asking this question. When I use a pretrained weight with imgsz = 640 for prediction, even though there are differences between the .pth and .onnx formats, the discrepancies are still within an acceptable range. However, if I use a fine-tuned model weight, the differences become more significant. For example, the confidence score difference, which was initially around 0.01 ~ 0.02, now becomes much larger—where the .pth model detects an object with 0.45 confidence, but the .onnx model fails to detect anything.

I want to understand the reason behind this issue. Is it because my default fine-tuning method is problematic, or is it due to differences in preprocessing between .onnx and .pth? If it's the latter, why doesn't this issue occur when using the pretrained weight?

**Comment by Y-T-G** [comment]
What if you export ONNX with `dynamic=True`?

**Comment by Ian-Work-AI** [comment]
@glenn-jocher @Y-T-G 
I think you might have misunderstood my point. I used the following code to make a prediction:

```
from ultralytics import YOLO

# Load a model
model = YOLO(<fine-tuned_weight>) 

model.predict(
        <img.png>,
        save = True, 
        imgsz = 640, 
        conf = 0.25
        )
```
But if the fine-tuned model I provide is in `.onnx` format, the results will be completely different from those of the `.pth` model, even with `dynamic=True`.

**Comment by Y-T-G** [comment]
Did you export with `dynamic=True`? Because if you do, then the preprocessing is the same as PyTorch.

**Comment by jj861119** [comment]
Hi @Y-T-G , I'm the co-worker of @Ian-Work-AI . Let me explain the problem we encountered more clearly.

We've trained .pth model with following CLI command
`yolo detect train model=yolov11n.pt epochs=100 imgsz=640`

Then, we exported the .pth model into two onnx models with following CLI commands.
(384x640 is result from 16:9 input image resolution)
A. `yolo export model="best.pt" format=onnx imgsz=[384,640]` -> A.onnx
B. `yolo export model="best.pt" format=onnx` -> B.onnx

Finally, we get detections with following CLI commands.
A. `yolo predict model="A.onnx" source="images.jpg" imgsz=[384,640]`
B. `yolo predict model="B.onnx" source="images.jpg"`

However, we get totally different results of A.onnx and B.onnx.
Furthermore, the result of A.onnx (384x640) is better than B.onnx (640x640) despite we trained the model with 640x640 resolution.

Result of A.onnx (384x640)
![Image](https://github.com/user-attachments/assets/77203615-1d35-4bdc-ad84-712ff27a34e1)

Result of B.onnx (640x640)
![Image](https://github.com/user-attachments/assets/4eae2227-d2b1-43dd-831a-e6de9c64d6b8)

**Comment by Y-T-G** [comment]
The training using `mosaic` augmentation, so it's not exactly the same as training with `640x640` because the image is filled with other images, not with padding. If you want to train with padding, then you need to use `mosaic=0` during training.

**Comment by jj861119** [comment]
I understand the difference of training augmentation and inference padding preprocessing.
But do you know what causes the different results of A.onnx and B.onnx?
As far as i know, YOLO model should be invariant of the object translation made by padding.

**Comment by glenn-jocher** [comment]
The difference arises from **input preprocessing alignment** between PyTorch and exported ONNX models. When you export with specific `imgsz` ([384,640] vs default [640,640]), the ONNX graph hardcodes these dimensions, altering scaling/padding operations compared to PyTorch's dynamic handling. 

For consistent results:
1. **Match training/imgsz**: Export ONNX with the **exact same `imgsz` used during training** (`640` in your case)
2. **Use `dynamic=True`** during export to maintain PyTorch-like aspect ratio handling:  
   `yolo export model=

**Comment by jj861119** [comment]
After testing on pytorch models, it still outputs totally different results between 384x640 and 640x640 inputs.
I think this is the same issue that @Y-T-G metioned in https://github.com/ultralytics/ultralytics/issues/14733#issuecomment-2254102334 : "The difference in padding approach can cause a difference in confidence."

Do you know the reason why padding approach makes difference in confidence?
Besides, what's the correct padding approach I should use for a better prediction result?

**Comment by Y-T-G** [comment]
What's the confidence with the original PyTorch model

**Comment by jj861119** [comment]
Same as the onnx model output above.

**Comment by Y-T-G** [comment]
The first one?

**Comment by jj861119** [comment]
The output of pytorch model using 384x640 input is same as the A.onnx (384x640) output image above.
The output of pytorch model using 640x640 input is same as the B.onnx (640x640) output image above.

**Comment by Y-T-G** [comment]
What command did you use to train

**Comment by jj861119** [comment]
yolo detect train model=yolov11n.pt epochs=100 imgsz=640

**Comment by Y-T-G** [comment]
What's the output if you use `iou=0.9` during prediction

**Comment by jj861119** [comment]
I got these results using iou=0.9.
384x640 :
![Image](https://github.com/user-attachments/assets/ac48b436-1cfe-4617-ac49-b8b4c938cdb1)
640x640 :
![Image](https://github.com/user-attachments/assets/8ad19727-bc61-4308-b216-7458aa3ddcd5)

What's the meaning of testing this?

**Comment by Y-T-G** [comment]
I wanted to see if the apples were detected but got removed due to NMS.

**Comment by Y-T-G** [comment]
> The training using `mosaic` augmentation, so it's not exactly the same as training with `640x640` because the image is filled with other images, not with padding. If you want to train with padding, then you need to use `mosaic=0` during training.

You should use this to train the model. Any changes to the image affects prediction. Even simply changing the interpolation method used for resizing would affect prediction

**Comment by jj861119** [comment]
We did use mosaic augmentation to train the model.
According to the documentation (https://docs.ultralytics.com/usage/cfg/#augmentation-settings), mosaic is set to 1.0 by default.

**Comment by Y-T-G** [comment]
Yeah. I want you to disable it. `mosaic=0` disables it.

**Comment by jj861119** [comment]
Setting mosaic=0 for training still results in different confidence score between 384x640 and 640x640 inputs.

384x640
![Image](https://github.com/user-attachments/assets/8af139e8-e015-43a1-8713-08a2534baaa4)
640x640
![Image](https://github.com/user-attachments/assets/4552892a-7bca-494b-82f0-aa5753c442d0)

**Comment by glenn-jocher** [comment]
The confidence differences stem from input aspect ratio changes affecting feature alignment in the model architecture. For consistent results:

1. **Maintain training aspect ratio** during inference (640x640)
2. Use **identical preprocessing** (resize strategy/padding) for both PyTorch and ONNX
3. Export with `imgsz=640` and `dynamic=True` to preserve aspect ratio handling

The model's sensitivity to input scaling is inherent to CNN architectures. For optimal results, use the same resolution and preprocessing pipeline throughout the training/inference cycle.

**Comment by glenn-jocher** [comment]
The pretrained models are trained with multiple aspect ratios, whereas your custom model likely isn't, resulting in greater sensitivity to input shape variations. You can train your custom model with rectangular training enabled by using `rect=True`.
